src/algebra/free_algebra.lean

```diff
@@ -0,0 +1,293 @@
+/-
+Copyright (c) 2020 Adam Topaz. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Scott Morrison, Adam Topaz.
+-/
+
+import ring_theory.algebra
+import linear_algebra
+import data.monoid_algebra
+
+/-!
+# Free Algebras
+
+Given a commutative semiring `R`, and a type `X`, we construct the free `R`-algebra on `X`.
+
+## Notation
+
+1. `free_algebra R X` is the free algebra itself. It is endowed with an `R`-algebra structure.
+2. `free_algebra.ι R X` is the function `X → free_algebra R X`.
+3. Given a function `f : X → A` to an R-algebra `A`, `lift R X f` is the lift of `f` to an
+  `R`-algebra morphism `free_algebra R X → A`.
+
+## Theorems
+
+1. `ι_comp_lift` states that the composition `(lift R X f) ∘ (ι R X)` is identical to `f`.
+2. `lift_unique` states that whenever an R-algebra morphism `g : free_algebra R X → A` is
+  given whose composition with `ι R X` is `f`, then one has `g = lift R X f`.
+3. `hom_ext` is a variant of `lift_unique` in the form of an extensionality theorem.
+4. `lift_comp_ι` is a combination of `ι_comp_lift` and `lift_unique`. It states that the lift
+  of the composition of an algebra morphism with `ι` is the algebra morphism itself.
+5. `equiv_monoid_algebra_free_monoid : free_algebra R X ≃ₐ[R] monoid_algebra R (free_monoid X)`
+
+## Implementation details
+
+As noted above, the free algebra on `X` is constructed as the free `R`-algebra generated by `X`.
+This is done as a quotient of an inductive type by an inductively defined relation.
+Explicitly, the construction involves three steps:
```

Comment:
```suggestion
The construction of the free algebra involves three steps:
```
Doesn't it feel silly to say that "the free algebra on X is constructed as the free R-algebra on X"?

Reply:
Or perhaps with forward references
```suggestion
We construct the free algebra on `X` as a quotient of an inductive type `free_algebra.pre` by an inductively defined relation `free_algebra.rel`.
Explicitly, the construction involves three steps:
```

Reply:
Oops!